File: src/lib/providers/seedance.js


import { buildVideoDirective } from "../video-directive";
import { parseRefRoles } from "../shot-spec";

/** Instant revert path: SEEDANCE_LEGACY_DIRECTIVE=1 restores the pre-2026-07-28
 *  hand-written directives on BOTH Seedance paths, without a deploy. The new
 *  wording is reasoned rather than bake-off measured (video comparisons cost
 *  real generations), so a one-env-var undo is the honest safety net. */
export function legacyDirective() {
  return process.env.SEEDANCE_LEGACY_DIRECTIVE === "1";
}

/** The previous directive, kept verbatim for that revert path only. Note the
 *  photoreal-only assumptions ("skin tone and texture", "never beautified")
 *  and the unconditional focus directive — the three faults video-directive.ts
 *  documents and fixes. */
function legacyHeroDirective(refCount) {
  if (!refCount) return "";
  return (
    `IDENTITY LOCK: the reference image${refCount > 1 ? "s" : ""} define ` +
    `the MAIN CHARACTER's exact, fixed appearance. In EVERY frame keep this ` +
    `exact same person — identical face (bone structure, jawline, hairline, ` +
    `eye shape and color, eyebrows, nose, lips, skin tone and texture with ` +
    `its moles/scars/freckles, facial hair, apparent age), plus the same ` +
    `hairstyle, body build and worn outfit/jewelry unless the prompt ` +
    `explicitly changes them — unmistakably the SAME individual, never a ` +
    `lookalike, never beautified or idealized, with zero identity or ` +
    `wardrobe drift between frames. Keep the main character in sharp ` +
    `foreground focus as the clear focal point. Every other person (crowd, ` +
    `bystanders, dancers, background figures) is a DIFFERENT anonymous ` +
    `individual who must NOT share or resemble the main character's face; ` +
    `render the crowd softer and out of focus so it never competes with or ` +
    `is mistaken for the main character. Never duplicate the main character. ` +
    `LITERAL PROMPT: execute the prompt exactly as written — every stated ` +
    `subject, count, wardrobe item, color, action, camera move and lighting ` +
    `appears precisely as specified; add nothing, drop nothing, reinterpret ` +
    `nothing. Anything under "NEGATIVE PROMPT" or phrased as "no …" is ` +
    `strictly forbidden in every frame. `
  );
}

/** Error carrying a machine-readable code so callers can branch (e.g. offer a
 *  text-to-video retry when a reference image is rejected by moderation). */
export class SeedanceError extends Error {
  code;
  status;
  constructor(message, code = "seedance_error", status) {
    super(message);
    this.name = "SeedanceError";
    this.code = code;
    this.status = status;
  }
}

function arkBase() {
  return (
    process.env.ARK_BASE_URL || "https://ark.ap-southeast.bytepluses.com/api/v3"
  ).replace(/\/$/, "");
}

function arkKey() {
  const key = process.env.ARK_API_KEY;
  if (!key) {
    throw new Error(
      "ARK_API_KEY is not set. Add it to .env.local (Seedance / BytePlus ModelArk)."
    );
  }
  return key;
}

const STANDARD_MODEL =
  process.env.SEEDANCE_MODEL || "dreamina-seedance-2-0-260128";
const FAST_MODEL =
  process.env.SEEDANCE_MODEL_FAST || "dreamina-seedance-2-0-fast-260128";
const MODEL_25 =
  process.env.SEEDANCE_MODEL_25 || "dreamina-seedance-2-5-260628";

function pickModel(modelDisplay) {
  // Checked before mini/fast/lite: "Seedance 2.5" doesn't contain any of
  // those words today, but a hypothetical "Seedance 2.5 Mini" shouldn't fall
  // through to the 2.0 fast SKU if one ever ships.
  if (modelDisplay && /2\.5/.test(modelDisplay)) return MODEL_25;
  if (modelDisplay && /\b(mini|fast|lite)\b/i.test(modelDisplay)) return FAST_MODEL;
  return STANDARD_MODEL;
}

/** Per-reference role hint for buildVideoDirective's legend (2026-08-17,
 *  video-directive.ts "PER-REFERENCE ROLE LEGEND"). `refs` is whatever was
 *  actually resolved onto this request (resolveReferences' tagged subset, or
 *  everything when the prompt tags nothing) — each entry already carries its
 *  own `@imgN` tag and 1-based original `index`, so this only has to look up
 *  each attached ref's role and key the map by that same original index,
 *  which is exactly the number `tagsToImageRefs` prints as "[image N]". No
 *  vision call: parseRefRoles is the same free keyword scan the image path
 *  already runs. Returns undefined (not an empty Map) when nothing is
 *  resolvable, so buildVideoDirective's own `refRoles` default takes over. */
function buildRefRoles(refs, rawPrompt) {
  if (!refs.length) return undefined;
  const roleByTag = parseRefRoles(rawPrompt);
  if (!roleByTag.size) return undefined;
  const map = new Map();
  for (const ref of refs) {
    const role = roleByTag.get(ref.tag);
    if (role) map.set(ref.index, role);
  }
  return map.size ? map : undefined;
}

/** Seedance reads "[image N]" references in the prompt. Translate the UI's
 *  @imgN tags so the model binds each tag to the matching reference_image. */
function tagsToImageRefs(prompt) {
  return prompt
    .replace(/@img(\d+)/gi, (_, n) => `[image ${n}]`)
    // Same convention for clips. Unlike the image form this one is NOT
    // probe-verified — reference clips are attached as content items and work
    // without any in-prompt token, so the worst case is the model reading this
    // as ordinary text rather than a broken request.
    .replace(/@vid(\d+)/gi, (_, n) => `[video ${n}]`);
}

export const MODERATION_MESSAGE =
  "BytePlus rejected the reference image — its privacy / anti-deepfake filter flags photorealistic faces (it can't tell an AI-generated face from a real one). Retry as text-to-video, or use a clearly stylized reference.";

/** Whether an error code/message looks like a BytePlus moderation rejection. */
export function isModerationMessage(text) {
  return /SensitiveContent|Privacy|real person|portrait|sensitive/i.test(
    text || ""
  );
}

/** Turn raw ModelArk error bodies into a typed, UI-friendly SeedanceError. */
function friendlyError(status, body) {
  let code = "";
  let message = "";
  try {
    const j = JSON.parse(body);
    code = j?.error?.code || "";
    message = j?.error?.message || "";
  } catch {
    /* not JSON */
  }
  if (isModerationMessage(code + message)) {
    return new SeedanceError(MODERATION_MESSAGE, "moderation", status);
  }
  if (code)
    return new SeedanceError(
      `Seedance error (${status} ${code}): ${message || body.slice(0, 300)}`,
      "seedance_error",
      status
    );
  return new SeedanceError(
    `Seedance create error ${status}: ${body.slice(0, 400)}`,
    "seedance_error",
    status
  );
}

/**
 * Minimal task-type trigger sentences for Seedance 2.5's Edit/Extend modes
 * (see the file header — BytePlus classifies by content role + these exact
 * kinds of phrases, not a request field). Deliberately NOT run through
 * video-directive.ts: that module's identity-lock/style-follow scaffolding is
 * built for GENERATING a new video from a reference, and this codebase
 * already learned once (video-directive.ts's own header) that stacking
 * unrelated directives contradicts rather than adds — feeding "keep the
 * subject in sharp foreground focus" etc. at an Edit task would compete with
 * the user's actual edit instructions for no benefit. The user's raw prompt
 * carries the instructions; this prefix only has to get BytePlus's
 * classifier to recognize which task type it's looking at.
 */
const EDIT_TRIGGER = "Edit the attached reference video as follows: ";
const EXTEND_TRIGGER = "Extend the attached reference video forward in time: ";

export async function createVideoTask(
  input
) {
  const model = pickModel(input.modelDisplay);
  const refs = input.references ?? [];
  const refRole = process.env.SEEDANCE_IMAGE_ROLE || "reference_image";
  const taskMode = input.taskMode ?? "generate";

  // Identity/style scaffolding now lives in lib/video-directive.ts, shared with
  // the Higgsfield path so the two cannot drift apart again. It also assembles
  // the whole text (scaffolding, prompt verbatim, then the precedence rule),
  // because the closing rule has to land AFTER the prompt — which the old
  // `directive + prompt` shape made impossible.
  //
  // Edit/Extend skip it entirely — see the trigger-sentence comment above.
  let text;
  if (taskMode === "edit") {
    text = EDIT_TRIGGER + tagsToImageRefs(input.prompt.trim());
  } else if (taskMode === "extend") {
    text = EXTEND_TRIGGER + tagsToImageRefs(input.prompt.trim());
  } else {
    text = legacyDirective()
      ? legacyHeroDirective(refs.length) + tagsToImageRefs(input.prompt.trim())
      : buildVideoDirective({
          prompt: tagsToImageRefs(input.prompt.trim()),
          refCount: refs.length,
          tagSyntax: "bracket",
          refRoles: buildRefRoles(refs, input.prompt),
        });
  }

  const content = [{ type: "text", text }];
  refs.forEach((ref) => {
    content.push({
      type: "image_url",
      image_url: { url: ref.dataUrl },
      role: refRole,
    });
  });
  // `role` is required here, unlike on image items — see the header. Capped at
  // the documented 3 rather than letting the provider reject the whole request.
  for (const url of (input.referenceVideoUrls ?? []).slice(0, 3)) {
    content.push({
      type: "video_url",
      video_url: { url },
      role: "reference_video",
    });
  }

  const body = {
    model,
    content,
    // Was hardcoded false. Still defaults to false when the caller says
    // nothing, so nothing starts paying for audio it did not ask for.
    generate_audio: input.generateAudio === true,
  };
  if (taskMode === "edit" || taskMode === "extend") {
    // BOTH task types require ratio:"adaptive" (output follows the source
    // clip's own aspect ratio) — sending the UI's own aspectRatio here would
    // 400 as InvalidParameter.TaskTypeConstraint, reported async on the next
    // poll rather than on this create call, because classification happens
    // after the task is already queued. Edit additionally requires
    // duration:-1 (output matches the source's length); Extend allows a real
    // duration if the caller gave one.
    body.ratio = "adaptive";
    body.duration = taskMode === "edit" ? -1 : input.duration || -1;
  } else {
    if (input.ratio) body.ratio = input.ratio;
    if (input.duration) body.duration = input.duration;
  }
  if (input.resolution) body.resolution = input.resolution;

  const res = await fetch(`${arkBase()}/contents/generations/tasks`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${arkKey()}`,
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const text = await res.text();
    throw friendlyError(res.status, text);
  }
  const json = await res.json();
  const id = json?.id || json?.task_id || json?.data?.id;
  if (!id) throw new Error("Seedance create: no task id in response.");
  return id;
}

export async function getVideoTask(
  taskId
) {
  const res = await fetch(
    `${arkBase()}/contents/generations/tasks/${encodeURIComponent(taskId)}`,
    {
      method: "GET",
      headers: { Authorization: `Bearer ${arkKey()}` },
    }
  );
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Seedance poll error ${res.status}: ${text.slice(0, 500)}`);
  }
  const json = await res.json();

  // ModelArk statuses: queued | running | succeeded | failed | cancelled
  const rawStatus = (json?.status || "").toLowerCase();
  let status = "running";
  if (rawStatus === "succeeded") status = "succeeded";
  else if (rawStatus === "failed" || rawStatus === "cancelled") status = "failed";
  else if (rawStatus === "queued") status = "queued";

  const videoUrl =
    json?.content?.video_url ||
    json?.content?.[0]?.video_url ||
    json?.video_url;

  const error =
    status === "failed"
      ? json?.error?.message || json?.error || "Generation failed"
      : undefined;

  // Seedance 2.5 only (2.0's response has no `usage` object) — see the file
  // header and pricing.ts computeSeedanceTokenCostCents.
  const totalTokensRaw = json?.usage?.total_tokens;
  const totalTokens =
    typeof totalTokensRaw === "number" && Number.isFinite(totalTokensRaw)
      ? totalTokensRaw
      : undefined;

  return { status, videoUrl, error, raw: json, totalTokens };
}

File: backend/apps/generation/providers/seedance.py
"""Port of src/lib/providers/seedance.js — Seedance 2.0/2.5 via BytePlus
ModelArk. See that file's header for the full probe-verified contract:
async create+poll, video-to-video content-item shape (role is MANDATORY,
unlike the image item), Edit/Extend's ratio:"adaptive" requirement, and
the "provider reports its own billing" pattern via usage.total_tokens.
"""

import os
import re

import requests

from ..video_directive import build_video_directive


def legacy_directive() -> bool:
    return os.environ.get("SEEDANCE_LEGACY_DIRECTIVE") == "1"


def _legacy_hero_directive(ref_count: int) -> str:
    if not ref_count:
        return ""
    plural = "s" if ref_count > 1 else ""
    return (
        f"IDENTITY LOCK: the reference image{plural} define "
        "the MAIN CHARACTER's exact, fixed appearance. In EVERY frame keep this "
        "exact same person — identical face (bone structure, jawline, hairline, "
        "eye shape and color, eyebrows, nose, lips, skin tone and texture with "
        "its moles/scars/freckles, facial hair, apparent age), plus the same "
        "hairstyle, body build and worn outfit/jewelry unless the prompt "
        "explicitly changes them — unmistakably the SAME individual, never a "
        "lookalike, never beautified or idealized, with zero identity or "
        "wardrobe drift between frames. Keep the main character in sharp "
        "foreground focus as the clear focal point. Every other person (crowd, "
        "bystanders, dancers, background figures) is a DIFFERENT anonymous "
        "individual who must NOT share or resemble the main character's face; "
        "render the crowd softer and out of focus so it never competes with or "
        "is mistaken for the main character. Never duplicate the main character. "
        'LITERAL PROMPT: execute the prompt exactly as written — every stated '
        "subject, count, wardrobe item, color, action, camera move and lighting "
        "appears precisely as specified; add nothing, drop nothing, reinterpret "
        'nothing. Anything under "NEGATIVE PROMPT" or phrased as "no …" is '
        "strictly forbidden in every frame. "
    )


class SeedanceError(Exception):
    def __init__(self, message: str, code: str = "seedance_error", status: int | None = None):
        super().__init__(message)
        self.code = code
        self.status = status


def _ark_base() -> str:
    return os.environ.get("ARK_BASE_URL", "https://ark.ap-southeast.bytepluses.com/api/v3").rstrip("/")


def _ark_key() -> str:
    key = os.environ.get("ARK_API_KEY")
    if not key:
        raise RuntimeError("ARK_API_KEY is not set. Add it to .env.local (Seedance / BytePlus ModelArk).")
    return key


def _standard_model() -> str:
    return os.environ.get("SEEDANCE_MODEL", "dreamina-seedance-2-0-260128")


def _fast_model() -> str:
    return os.environ.get("SEEDANCE_MODEL_FAST", "dreamina-seedance-2-0-fast-260128")


def _model_25() -> str:
    return os.environ.get("SEEDANCE_MODEL_25", "dreamina-seedance-2-5-260628")


def _pick_model(model_display: str | None) -> str:
    if model_display and re.search(r"2\.5", model_display):
        return _model_25()
    if model_display and re.search(r"\b(mini|fast|lite)\b", model_display, re.IGNORECASE):
        return _fast_model()
    return _standard_model()


def _tags_to_image_refs(prompt: str) -> str:
    prompt = re.sub(r"@img(\d+)", lambda m: f"[image {m.group(1)}]", prompt, flags=re.IGNORECASE)
    prompt = re.sub(r"@vid(\d+)", lambda m: f"[video {m.group(1)}]", prompt, flags=re.IGNORECASE)
    return prompt


MODERATION_MESSAGE = (
    "BytePlus rejected the reference image — its privacy / anti-deepfake filter flags photorealistic "
    "faces (it can't tell an AI-generated face from a real one). Retry as text-to-video, or use a "
    "clearly stylized reference."
)


def is_moderation_message(text: str) -> bool:
    return bool(re.search(r"SensitiveContent|Privacy|real person|portrait|sensitive", text or "", re.IGNORECASE))


def _friendly_error(status: int, body: str) -> SeedanceError:
    code = ""
    message = ""
    try:
        import json as _json

        j = _json.loads(body)
        code = (j.get("error") or {}).get("code") or ""
        message = (j.get("error") or {}).get("message") or ""
    except ValueError:
        pass
    if is_moderation_message(code + message):
        return SeedanceError(MODERATION_MESSAGE, "moderation", status)
    if code:
        return SeedanceError(f"Seedance error ({status} {code}): {message or body[:300]}", "seedance_error", status)
    return SeedanceError(f"Seedance create error {status}: {body[:400]}", "seedance_error", status)


EDIT_TRIGGER = "Edit the attached reference video as follows: "
EXTEND_TRIGGER = "Extend the attached reference video forward in time: "


def create_video_task(
    prompt: str, model_display: str | None = None, ratio: str | None = None, resolution: str | None = None,
    duration: int | None = None, references: list[dict] | None = None, reference_video_urls: list[str] | None = None,
    generate_audio: bool = False, task_mode: str = "generate", seed: int | None = None,
) -> str:
    """references: [{"dataUrl": "..."}] (already resolved LabeledRef dicts,
    only .dataUrl is used here). Returns the created task id."""
    model = _pick_model(model_display)
    refs = references or []
    ref_role = os.environ.get("SEEDANCE_IMAGE_ROLE", "reference_image")

    if task_mode == "edit":
        text = EDIT_TRIGGER + _tags_to_image_refs(prompt.strip())
    elif task_mode == "extend":
        text = EXTEND_TRIGGER + _tags_to_image_refs(prompt.strip())
    else:
        tagged_prompt = _tags_to_image_refs(prompt.strip())
        text = (
            _legacy_hero_directive(len(refs)) + tagged_prompt
            if legacy_directive()
            else build_video_directive(tagged_prompt, len(refs), "bracket")
        )

    content: list[dict] = [{"type": "text", "text": text}]
    for ref in refs:
        content.append({"type": "image_url", "image_url": {"url": ref["dataUrl"]}, "role": ref_role})
    for url in (reference_video_urls or [])[:3]:
        content.append({"type": "video_url", "video_url": {"url": url}, "role": "reference_video"})

    body: dict = {"model": model, "content": content, "generate_audio": generate_audio is True}
    if task_mode in ("edit", "extend"):
        body["ratio"] = "adaptive"
        body["duration"] = -1 if task_mode == "edit" else (duration or -1)
    else:
        if ratio:
            body["ratio"] = ratio
        if duration:
            body["duration"] = duration
    if resolution:
        body["resolution"] = resolution
    # Real, documented ModelArk field (Phase 3.1) — mirrors seedance.js's
    # identical convention, omitted entirely when the caller has no seed.
    if isinstance(seed, int):
        body["seed"] = seed

    res = requests.post(
        f"{_ark_base()}/contents/generations/tasks",
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {_ark_key()}"},
        json=body,
        timeout=30,
    )
    if not res.ok:
        raise _friendly_error(res.status_code, res.text)
    data = res.json()
    task_id = data.get("id") or data.get("task_id") or (data.get("data") or {}).get("id")
    if not task_id:
        raise RuntimeError("Seedance create: no task id in response.")
    return task_id


def get_video_task(task_id: str) -> dict:
    from urllib.parse import quote

    res = requests.get(
        f"{_ark_base()}/contents/generations/tasks/{quote(task_id)}",
        headers={"Authorization": f"Bearer {_ark_key()}"},
        timeout=30,
    )
    if not res.ok:
        raise RuntimeError(f"Seedance poll error {res.status_code}: {res.text[:500]}")
    data = res.json()

    raw_status = (data.get("status") or "").lower()
    if raw_status == "succeeded":
        status = "succeeded"
    elif raw_status in ("failed", "cancelled"):
        status = "failed"
    elif raw_status == "queued":
        status = "queued"
    else:
        status = "running"

    content = data.get("content")
    video_url = None
    if isinstance(content, dict):
        video_url = content.get("video_url")
    elif isinstance(content, list) and content:
        video_url = content[0].get("video_url")
    video_url = video_url or data.get("video_url")

    error = None
    if status == "failed":
        err = data.get("error")
        error = (err.get("message") if isinstance(err, dict) else err) or "Generation failed"

    total_tokens_raw = (data.get("usage") or {}).get("total_tokens")
    total_tokens = total_tokens_raw if isinstance(total_tokens_raw, (int, float)) else None

    return {"status": status, "videoUrl": video_url, "error": error, "raw": data, "totalTokens": total_tokens}
